src/NVMLBoardUtilizationAgent.cpp

```diff
@@ -0,0 +1,294 @@
+                }
+
+                // set Xeon frequency control
+                freq_ctl_itr = m_control_available.find("FREQUENCY");
+                if (xeon_freq_request != freq_ctl_itr->second.m_last_setting) {
+                    m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, xeon_freq_request);
+                    freq_ctl_itr->second.m_last_setting = xeon_freq_request;
+                }
+
+                m_do_write_batch = true;
+            }
+        }
+    }
+
+    //If new values have been adjusted, write
+    bool NVMLBoardUtilizationAgent::do_write_batch(void) const
+    {
+        return m_do_write_batch;
+    }
+
+    // Read signals from the platform and calculate samples to be sent up
+    void NVMLBoardUtilizationAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        assert(out_sample.size() == M_NUM_SAMPLE);
+        // Collect latest signal values
+        for (auto &sv : m_signal_available) {
+            sv.second.m_last_signal = m_platform_io.sample(sv.second.m_batch_idx);
+        }
+    }
+
+    // Wait for the remaining cycle time to keep Controller loop cadence at 1 second
+    void NVMLBoardUtilizationAgent::wait(void)
+    {
+        geopm_time_s current_time;
+        do {
+            geopm_time(&current_time);
+        }
+        while(geopm_time_diff(&m_last_wait, &current_time) < M_WAIT_SEC);
+        geopm_time(&m_last_wait);
+    }
+
+    // Adds the wait time to the top of the report
+    std::vector<std::pair<std::string, std::string> > NVMLBoardUtilizationAgent::report_header(void) const
+    {
+        return {{"Wait time (sec)", std::to_string(M_WAIT_SEC)}};
+    }
+
+    // Adds to the per-node section of the report
+    std::vector<std::pair<std::string, std::string> > NVMLBoardUtilizationAgent::report_host(void) const
+    {
+        return {
+            {"Accelerator Frequency Requests", std::to_string(m_accelerator_frequency_requests)},
+        };
+    }
+
+    // This Agent does not add any per-region details
+    std::map<uint64_t, std::vector<std::pair<std::string, std::string> > > NVMLBoardUtilizationAgent::report_region(void) const
+    {
+        return {};
+    }
+
+    // Adds trace columns samples and signals of interest
+    std::vector<std::string> NVMLBoardUtilizationAgent::trace_names(void) const
+    {
+        std::vector<std::string> names;
+        // Build name in the format: "NVML::FREQUENCY-board-0"
+        for (auto &sv : m_signal_available) {
+            names.push_back(sv.first + "-" + m_platform_topo.domain_type_to_name(GEOPM_DOMAIN_BOARD) + "-0");
+        }
+        return names;
+    }
+
+    // Updates the trace with values for samples and signals from this Agent
+    void NVMLBoardUtilizationAgent::trace_values(std::vector<double> &values)
+    {
+        int values_idx = 0;
+        // Default assumption is that every signal added should be in the trace
+        for (auto &sv : m_signal_available) {
```

Comment:
Kind of a nit: this could be regular for(values_idx = 0.. N) type loop assuming the vectors are the same size.